File: python/runcached.py
#! /usr/bin/env -S python3 -O -u
# -*- Mode: Python; py-indent-offset: 4; coding: utf-8 -*-
# vim: fenc=utf-8 tabstop=4 softtabstop=4 shiftwidth=4 expandtab

# runcached
# Execute commands while caching their output for subsequent calls. 
# Command output will be cached for $cacheperiod and replayed for subsequent calls
#
# 2012
# Author Spiros Ioannou sivann <at> gmail.com
#
# 2024
# Author Pavel Vitis <pavelvitis@gmail.com>

from __future__ import annotations

__authors__ = [
    "Spiros Ioannou sivann <at> gmail.com",
    "Pavel Vitiš <pavelvitis@gmail.com>",
]
__maintainer__ = "Pavel Vitis <pavelvitis@gmail.com"
__license__ = "Apache License, Version 2.0"
__status__ = "Development"

import argparse
import fcntl
import os
import random
import signal
import subprocess
import sys
import time
from collections.abc import Sequence, Callable
from contextlib import suppress, contextmanager
from hashlib import md5
from pathlib import Path
from shutil import copyfileobj
from subprocess import Popen
from typing import Iterable, Optional

# Avoid importing psutil if not necessary
if Path("/proc").is_dir():
    pid_exists = lambda pid: Path(f"/proc/{pid}").exists()
else:
    from psutil import pid_exists

# Configurable parameters
DEFAULT_CACHE_TIMEOUT_SEC: float = 20
MAX_WAIT_PREV_SEC: int = 5
MIN_RAND_SEC: float = 0
MAX_RAND_SEC: float = 0

os.environ['PYTHONUNBUFFERED'] = '1'


def get_cache_dir() -> Path:
    """
    Return the system's caching directory.
    """
    from tempfile import gettempdir
    return Path(gettempdir())


def generate_command_hash(command: Iterable[str]) -> str:
    """
    Generate an MD5 hash for the given command.
    """
    # noinspection InsecureHash
    return md5(" ".join(command).encode("utf-8")).hexdigest()


class CommandCache(object):
    def __init__(self, command:Sequence[str], cache_timeout:float):
        self.cache_timeout = cache_timeout
        self.command = command
        self.cache_dir = get_cache_dir()
        self.command_hash = generate_command_hash(command)
        self.output_cache = Path(self.cache_dir, f"{self.command_hash}.data")
        self.exit_file = Path(self.cache_dir, f"{self.command_hash}.exit")
        self.cmd_file = Path(self.cache_dir, f"{self.command_hash}.cmd")
        if not self.cmd_file.is_file():
            with self.cmd_file.open('w') as f:
                f.write(" ".join(self.command))

    def is_valid(self, cache_timeout: Optional[float] = None):
        if cache_timeout is None:
            cache_timeout = self.cache_timeout
        return self.output_cache.is_file() and time.time() - self.output_cache.stat().st_mtime <= cache_timeout

    def invalidate(self):
        self.output_cache.unlink()

    def cache_result(self, f: Iterable[str]):
        output_cache_file_encoding = "utf-8"
        with self.output_cache.open('w', encoding=output_cache_file_encoding) as f_cache:
            try:
                while True:
                    line = next(f)
                    f_cache.write(line)
                    f_cache.flush()
            except StopIteration as s:
                return_code = s.value
                # print(f"{return_code=}")
            except BrokenPipeError:
                return_code = 0
                # print(f"{return_code=}")
            except KeyboardInterrupt:
                return_code = 0
                # print(f"{return_code=}")
                raise
            finally:
                with self.exit_file.open('w') as f:
                    f.write(str(return_code))

                # Must update the modification timestamp so that the command runtime does not add to cache expiration timeout
                self.output_cache.touch()

        return return_code


def execute_command(command: Sequence[str], output_cache_file_encoding: str = "utf-8") -> Iterable[str]:
    """
    Execute a command and redirect results into cache files:

        - execution return_code into exit_file
        - stdout & stderr into output_cache_file
    """
    return_code = 1
    keyboard_interrupt = False
    process = Popen(command, stdout=subprocess.PIPE, stderr=subprocess.DEVNULL)
    while process.poll() is None:
        try:
            if True: #process.stdout.readable():
                # print("Reading the line from command output...", file=sys.stderr)
                line = process.stdout.readline().decode(output_cache_file_encoding)
                try:
                    process.stdout.flush()
                    if line:
                        print(line, end="", file=sys.stdout, flush=True)
                except BrokenPipeError:
                    # print("BrokenPipeError...", file=sys.stderr)
                    devnull = os.open(os.devnull, os.O_WRONLY)
                    os.dup2(devnull, sys.stdout.fileno())
                    process.send_signal(signal.SIGPIPE)
                    return_code = 0
                    break
                else:
                    yield line
            else:
                break
                # time.sleep(.5)
        except KeyboardInterrupt:
            return_code = 1
            keyboard_interrupt = True
            break
    else:
        return_code = process.returncode

    if keyboard_interrupt:
        raise KeyboardInterrupt()
    else:
        return return_code


def wait_for_previous_command(pid_file: Path, wait_time_sec: int) -> bool:
    """
    Wait for a previous instance of the command to finish by checking for the existence
    of the given pid_file.

    Returns True if the pid_file disappears within a time limit, False on timeout.

    Also handles stale pid_file by checking if the process is still running. If it's not running,
    the pid_file is removed and function returns True.

    :returns: Return True on success, False on timeout
    """
    if not pid_file.is_file():
        # PID file cleaned up, process finished
        return True

    for _ in range(wait_time_sec):
        time.sleep(1)
        if not pid_file.is_file():
            # PID file cleaned up, process finished
            return True
        # Check for stale PID file
        try:
            pid: int = int(pid_file.read_text())
        except (ValueError, IOError):
            # Corrupted or not readable PID file
            # Try to remove it anyway
            with suppress(IOError):
                pid_file.unlink()
            return False
        else:
            if not pid_exists(pid):
                # Stale PID file, remove it
                pid_file.unlink()
                return True
    else:
        print("ERROR: Timeout waiting for previous command to finish.")
        return False


def send_text_to_stdout(text_file: Path) -> bool:
    """
    Send text file to stdout.

    Broken pipe error is handled by sending the rest of the text to /dev/null.

    See for more details:
    https://docs.python.org/3/library/signal.html#note-on-sigpipe

    :param text_file:
    :return: Return True on complete output, False on BrokenPipeError
    """
    try:
        # Output cached data
        # Open in binary mode and copy to stdout buffer directly to avoid unnecessary decoding/encoding
        with text_file.open('rb') as f:
            copyfileobj(f, sys.stdout.buffer)
        # Flush output here to force SIGPIPE to be triggered while inside this try block.
        sys.stdout.flush()
        return True
    except BrokenPipeError:
        # This handles BrokenPipeError on piping the result to 'head' or similar tools.
        # https://docs.python.org/3/library/signal.html#note-on-sigpipe
        # Python flushes standard streams on exit; redirect remaining output to devnull to avoid another
        # BrokenPipeError at shutdown
        devnull = os.open(os.devnull, os.O_WRONLY)
        os.dup2(devnull, sys.stdout.fileno())
        # Python exits with error code 1 on EPIPE
        return False


def create_pid_file(command: Iterable[str]) -> Optional[Path]:
    """
    Create PID file for given command.

    If there is a process with the same command already runhing, wait for its completion.

    Return None on timeout waiting for already running process.

    :param command:
    :return: PID if PID file has been successfully created or None if there was a timeout waiting for already running process
    """
    cache_dir = get_cache_dir()
    command_hash = generate_command_hash(command)
    pid_file = Path(cache_dir, f"{command_hash}.pid")

    # Random sleep
    if MAX_RAND_SEC - MIN_RAND_SEC > 0:
        time.sleep(random.uniform(MIN_RAND_SEC, MAX_RAND_SEC))

    # Avoid parallel execution
    if not wait_for_previous_command(pid_file, MAX_WAIT_PREV_SEC):
        return None

    # Create a PID file
    with pid_file.open('w') as f:
        f.write(str(os.getpid()))

    return pid_file


def main():
    """
    Run command and cache its output. Return cached output if cache not expired.
    """

    parser = argparse.ArgumentParser(description=main.__doc__)
    parser.add_argument("command", metavar="command ...", help="Command with arguments")
    parser.add_argument("-c", "--cache-timeout", type=float,
                        help=f"Cache timeout in seconds (float), default is {DEFAULT_CACHE_TIMEOUT_SEC}s")
    parser.add_argument("-e", "--cache-on-error", default=False, action="store_true",
                        help="Cache the command result also if it returns nonzero error code")
    parser.add_argument("-a", "--cache-on-abort", default=False, action="store_true",
                        help="Cache the command result also on ^C keyboard interrupt")
    parser.add_argument("-v", "--verbose", default=False, action="store_true",
                        help="Print diagnostic information")
    parser.add_argument('command_args', help=argparse.SUPPRESS, nargs=argparse.REMAINDER)
    args = parser.parse_args()

    command = [args.command] + args.command_args
    cache_timeout = max(0.0, args.cache_timeout if args.cache_timeout is not None else DEFAULT_CACHE_TIMEOUT_SEC)

    pid_file: Path = create_pid_file(command)
    # Timeout waiting for already running process
    try:
        if pid_file is None:
            print(f"ERROR: Process for given command still running: timeout ({MAX_WAIT_PREV_SEC}")
            sys.exit(2)
        if os.getpid() != int(pid_file.read_text()):
            print(f"ERROR: Creating PID file failed: Current process PID does not equal PID in the file.")
            sys.exit(2)
    except ValueError as e:
        print(f"ERROR: Creating PID file failed: {repr(e)}")
        sys.exit(2)
    except IOError as e:
        print(f"ERROR: Creating PID file failed: {repr(e)}")
        sys.exit(2)

    try:
        cache = CommandCache(command, cache_timeout=cache_timeout)

        # Execute and cache the result
        # Output cache file encoding must be the same as sys.stdout encoding
        cache_file_encoding = sys.stdout.encoding
        if not cache_file_encoding:
            cache_file_encoding = "utf-8"

        # If cache is still valid, return cached result
        if cache.is_valid(cache_timeout):
            if args.verbose:
                print("DIAG: Returning cached result")
            stdout_ok = send_text_to_stdout(cache.output_cache)
            # TODO: (pavel) 13/12/2024 Provide correct value
            return_code = int(cache.exit_file.read_text())
        else:
            # Check cache and execute the command if needed
            if args.verbose:
                print("DIAG: Executing command")
            executor = execute_command(command, output_cache_file_encoding=cache_file_encoding)
            return_code = cache.cache_result(executor)

        if return_code !=0 and not args.cache_on_error:
            with suppress(IOError):
                if args.verbose:
                    print("DIAG: Destroying output cache (RC)")
                cache.invalidate()

        # TODO: (pavel) 13/12/2024 Fix stdout_ok
        if True:
        # if stdout_ok:
           sys.exit(return_code)
        else:
            # On pipe-broken error
            sys.exit(1)
    except (KeyboardInterrupt, Exception) as e:
            # Cleanup on error and remove cache files
            # print(isinstance(e, KeyboardInterrupt))
            # print(f"{args.cache_on_error=}; {args.cache_on_abort=}")
            if args.cache_on_error or isinstance(e, KeyboardInterrupt) and args.cache_on_abort:
                raise
            if args.verbose:
                print("DIAG: Destroying output cache (KI)")
            with suppress(IOError):
                cache.invalidate()
            raise
    finally:
        # Cleanup the PID file
        if pid_file.is_file():
            pid_file.unlink()


if __name__ == "__main__":
    try:
        main()
    except (KeyboardInterrupt) as e:
        # raise
        ...
        # Suppress traceback on specific interrupts
        # print(e)

File: go/runcached.go
//
// runcached.go
// Copyright (C) 2024 Pavel Vitis <pavelvitis@gmail.com>
//
// Distributed under terms of the MIT license.
//

package main

import (
	"bufio"
	"bytes"
	"crypto/md5"
	"encoding/hex"
	"errors"
	"flag"
	"fmt"
	"io"
	"io/ioutil"
	"log"
	"math/rand"
	"os"
	"os/exec"
	"path/filepath"
	"strconv"
	"strings"
	"sync"
	"syscall"
	"time"
	"unicode"
)

const (
	defaultCacheTimeoutSec = 20.0
	maxWaitPrevSec         = 5
	minRandSec             = 0
	maxRandSec             = 0
)

var verbose bool
var debug bool

type CommandCache struct {
	Command      []string
	CacheTimeout float64
	CacheDir     string
	CommandHash  string
	OutputCache  string
	ExitFile     string
	CmdFile      string
}

func getCacheDir() string {
	tmpDir := os.TempDir()
	return tmpDir
}

func GenerateCommandHash(command []string) string {
	h := md5.New()
	h.Write([]byte(strings.Join(command, " ")))
	return hex.EncodeToString(h.Sum(nil))
}

// NewCommandCache creates a new CommandCache instance with the given command and cache timeout.
func NewCommandCache(command []string, cacheTimeout float64) *CommandCache {
	cacheDir := getCacheDir()
	commandHash := GenerateCommandHash(command)
	return &CommandCache{
		Command:      command,
		CacheTimeout: cacheTimeout,
		CacheDir:     cacheDir,
		CommandHash:  commandHash,
		OutputCache:  filepath.Join(cacheDir, fmt.Sprintf("%s.data", commandHash)),
		ExitFile:     filepath.Join(cacheDir, fmt.Sprintf("%s.exit", commandHash)),
		CmdFile:      filepath.Join(cacheDir, fmt.Sprintf("%s.cmd", commandHash)),
	}
}

// IsValid checks if the cached output is still valid based on the cache timeout.
func (cc *CommandCache) IsValid() bool {
    info, err := os.Stat(cc.OutputCache)
    if err != nil {
        return false
    }
	return time.Since(info.ModTime()).Seconds() <= cc.CacheTimeout
}

func (cc *CommandCache) Invalidate() {
	os.Remove(cc.OutputCache)
}

func (cc *CommandCache) CacheResult(output io.Reader) int {
	f, err := os.Create(cc.OutputCache)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to create cache file: %v\n", err)
		return 1
	}
	defer f.Close()

	scanner := bufio.NewScanner(output)
	var returnCode int
	for scanner.Scan() {
		line := scanner.Text()
		_, err := f.WriteString(line + "\n")
		if err != nil {
			fmt.Fprintf(os.Stderr, "Failed to write to cache file: %v\n", err)
			returnCode = 1
			break
		}
		fmt.Println(line)
	}
	if err := scanner.Err(); err != nil {
		fmt.Fprintf(os.Stderr, "Error reading command output: %v\n", err)
		returnCode = 1
	}

	exitFile, err := os.Create(cc.ExitFile)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to create exit file: %v\n", err)
		return 1
	}
	defer exitFile.Close()

	_, err = exitFile.WriteString(strconv.Itoa(returnCode))
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to write exit code to file: %v\n", err)
		return 1
	}

	return returnCode
}

func CustomParse(input string) ([]string, error) {
	var result []string
	var token strings.Builder
	var inQuotes bool
	var quoteChar rune

	for _, r := range input {
		switch {
		case inQuotes:
			if r == quoteChar { // End of quoted segment
				token.WriteRune(r)
				inQuotes = false
			} else {
				token.WriteRune(r)
			}
		case r == '"' || r == '\'':
			inQuotes = true
			quoteChar = r
			token.WriteRune(r) // Keep the opening quote
		case unicode.IsSpace(r):
			if token.Len() > 0 {
				result = append(result, token.String())
				token.Reset()
			}
		default:
			token.WriteRune(r)
		}
	}

	// Add the last token if it exists
	if token.Len() > 0 {
		result = append(result, token.String())
	}

	if inQuotes {
		return nil, fmt.Errorf("unclosed quote: %q", quoteChar)
	}

	return result, nil
}

func ConstructBashCommand(args []string) string {
	// Use shellwords to properly escape and join arguments
	escapedArgs, err := CustomParse(strings.Join(args, ` `))
	if err != nil {
		log.Fatal("Error parsing arguments:", err)
	}

	// Construct the final bash command with all arguments
// 	return "\"" + strings.Replace(strings.Join(escapedArgs, " "), "\"", "\\\"", -1) + "\""
	argsStr := strings.Join(escapedArgs, ` `)
//	argsStr = strings.Replace(argsStr, "\\", "\\\\", -1)
	argsStr = strings.Replace(argsStr, "\"", "\\\"", -1)
	return argsStr
}

func ExecuteCommand(command []string, cache *CommandCache) int {
    command = []string{"bash", "-i", "-c", ConstructBashCommand(command)}
//     log.Fatal(command)
	cmd := exec.Command(command[0], command[1:]...)
	output := &bytes.Buffer{}
	//cmd.Stdout = output
	cmd.Stderr = os.Stderr

	// Create a pipe to capture the output of the command
	stdoutPipe, err := cmd.StdoutPipe()
	if err != nil {
		fmt.Println("Error creating pipe:", err)
		return 1
	}

    // Open the cache file for writing
    cacheFile, err := os.Create(cache.OutputCache)
    if err != nil {
        fmt.Fprintf(os.Stderr, "Failed to create cache file: %v\n", err)
        return 1
    }
    defer cacheFile.Close()

    // Combine os.Stdout and cacheFile using io.MultiWriter
    multiWriter := io.MultiWriter(os.Stdout, cacheFile)

	err = cmd.Start()
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to start command: %v\n", err)
		return 1
	}

	// Create a wait group to synchronize the Goroutine
	var wg sync.WaitGroup
	wg.Add(1)

    // Use TeeReader to copy the output to both stdout and the buffer
	teeReader := io.TeeReader(stdoutPipe, output)

	// Copy data from the teeReader to os.Stdout on the fly
	go func() {
		defer wg.Done() // Decrement the counter when the Goroutine completes

        _, err := io.Copy(multiWriter, teeReader)
        if err != nil {
            fmt.Println("Error copying to writers:", err)
        }
	}()

	err = cmd.Wait()
	if err != nil {
		if exitErr, ok := err.(*exec.ExitError); ok {
            // Wait for the Goroutine to finish copying the data
            wg.Wait()

            exitFile, err := os.Create(cache.ExitFile)
            if err != nil {
                fmt.Fprintf(os.Stderr, "Failed to create exit file: %v\n", err)
                return 1
            }
            defer exitFile.Close()

            _, err = exitFile.WriteString(strconv.Itoa(exitErr.ExitCode()))
            if err != nil {
                fmt.Fprintf(os.Stderr, "Failed to write exit code to file: %v\n", err)
                return 1
            }

			return exitErr.ExitCode()
		}
		return 1
	}

	// Wait for the Goroutine to finish copying the data
    wg.Wait()

	return 0
}

func getPidFromFile(fileName string) int {
    data, err := ioutil.ReadFile(fileName)
    if err != nil {
        return -1
    }
    pid, err := strconv.Atoi(strings.TrimSpace(string(data)))
    if err != nil {
        return -1
    }
    return pid
}

func isProcessRunning(pid int) bool {
	// Find the process
	process, err := os.FindProcess(pid)
	if err != nil {
		// Process does not exist
		return false
	}

	// Try to send signal 0 to check if the process is running
	err = process.Signal(syscall.Signal(0))
	if err != nil {
		// If the error is syscall.ESRCH, the process is not running ("no such process")
		if errors.Is(err, syscall.ESRCH) {
			return false
		}
		// Otherwise, the process might not be accessible but is still running
		return true
	}

	return true
}

func waitForPreviousCommand(pidFile string, waitTimeSec int) bool {
	for i := 0; i < waitTimeSec; i++ {
		if _, err := os.Stat(pidFile); os.IsNotExist(err) {
			return true
		}
		pid := getPidFromFile(pidFile)
		// Can't read pid file
		if pid == -1 {
			os.Remove(pidFile)
			return false
		}
		// Stale pid file
		if !isProcessRunning(pid) {
			os.Remove(pidFile)
			return false
		}
		time.Sleep(1 * time.Second)
	}
	return false
}

func CreatePidFile(command []string) string {
	cacheDir := getCacheDir()
	commandHash := GenerateCommandHash(command)
	pidFile := filepath.Join(cacheDir, fmt.Sprintf("%s.pid", commandHash))

	if maxRandSec-minRandSec > 0 {
		time.Sleep(time.Duration(rand.Intn(maxRandSec-minRandSec)+minRandSec) * time.Second)
	}

	if !waitForPreviousCommand(pidFile, maxWaitPrevSec) {
		return ""
	}

	f, err := os.Create(pidFile)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to create PID file: %v\n", err)
		return ""
	}
	defer f.Close()

	_, err = f.WriteString(strconv.Itoa(os.Getpid()))
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to write PID to file: %v\n", err)
		return ""
	}

	return pidFile
}

func sendTextToStdout(textFile string) bool {
	f, err := os.Open(textFile)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to open file: %v\n", err)
		return false
	}
	defer f.Close()

	_, err = io.Copy(os.Stdout, f)
	if err != nil && err != io.EOF {
		fmt.Fprintf(os.Stderr, "Error copying file to stdout: %v\n", err)
		return false
	}

	return true
}

func readFileOrDefault(path string, defaultValue string) string {
	data, err := ioutil.ReadFile(path)
	if err != nil {
		return defaultValue
	}
	return string(data)
}

func RunCached(command []string, cacheTimeout float64, cacheOnError bool, cacheOnAbort bool) int {
	pidFile := CreatePidFile(command)
	if pidFile == "" {
		fmt.Fprintln(os.Stderr, "Timeout waiting for previous command to finish")
		return 2
	}
	defer os.Remove(pidFile)

	cache := NewCommandCache(command, cacheTimeout)
	if cache.IsValid() {
		if verbose {
			fmt.Println("Returning cached result")
		}
		sendTextToStdout(cache.OutputCache)
		returnCode, _ := strconv.Atoi(string(readFileOrDefault(cache.ExitFile, "0")))
		return returnCode
	} else {
		if verbose {
			fmt.Println("Executing command")
		}
		returnCode := ExecuteCommand(command, cache)
//		cache.CacheResult(output)

		if returnCode != 0 && !cacheOnError {
			cache.Invalidate()
		}
		return returnCode
	}
}

func printUsage() {
    fmt.Fprintln(os.Stderr, "Usage: runcached [-h] [-c CACHE_TIMEOUT] [-e] [-a] [-v] command ...")
}

func printHelp() {
    text := `Run command and cache its output. Return cached output if cache not expired.

positional arguments:
  command ...           Command with arguments

options:
  -h, --help            show this help message and exit
  -c CACHE_TIMEOUT, --cache-timeout CACHE_TIMEOUT
                        Cache timeout in seconds (float), default is 20s
  -e, --cache-on-error  Cache the command result also if it returns nonzero error code
  -a, --cache-on-abort  Cache the command result also on ^C keyboard interrupt
  -d, --debug           Debugging cache information
  -v, --verbose         Print diagnostic information
 `
    fmt.Fprint(os.Stderr, text)
}

// Function to manually escape quotes in arguments
func PreprocessArguments(args []string) []string {
	var processedArgs []string
	for _, arg := range args {
		// Escape backslashes if any
		arg = strings.ReplaceAll(arg, `\`, `\\`)
		// Preserve double-quotes by escaping them
		arg = strings.ReplaceAll(arg, `"`, `\"`)
		// Preserve quotes by escaping them
		if len(arg) >= 2 && arg[0] == '\'' && arg[len(arg)-1] == '\'' {
			arg = "\\'" + arg[1 : len(arg)-1] + "\\'"
		}
		processedArgs = append(processedArgs, arg)
	}
	return processedArgs
}

func main() {
// 	var cacheTimeout float64
	var cacheOnError, cacheOnAbort bool

    helpFlag := flag.Bool("help", false, "Show help")
	flag.BoolVar(helpFlag, "h", false, "Show help")
	cacheTimeoutFlag := flag.Float64("c", defaultCacheTimeoutSec, "Cache timeout in seconds")
	flag.Float64Var(cacheTimeoutFlag, "cache-timeout", defaultCacheTimeoutSec, "Cache timeout in seconds")
	flag.BoolVar(&cacheOnError, "e", false, "Cache command result even on nonzero error code")
	flag.BoolVar(&cacheOnAbort, "a", false, "Cache command result even on keyboard interrupt")
	flag.BoolVar(&debug, "d", false, "Enable cache debugging output")
	flag.BoolVar(&verbose, "v", false, "Enable verbose mode")
	args := os.Args[1 : len(os.Args)]
	flag.CommandLine.Parse(PreprocessArguments(args))

    if *helpFlag {
	    printUsage()
	    printHelp()
		os.Exit(1)
    }

	if flag.NArg() < 1 {
	    printUsage()
		os.Exit(1)
	}


//     log.Fatal(os.Args)
//     log.Fatal(flag.Args())
	command := flag.Args()

    os.Exit(RunCached(command, *cacheTimeoutFlag, cacheOnError, cacheOnAbort))
}
